### ISSUE TITLE
0.13.0 creates incorrect importmap entries

### ISSUE THREAD

**Issue by krishicks** [bazelbuild/bazel-gazelle#269]
I tried upgrading a project from rules_go 0.10.1 and gazelle 0.10.0 to 0.13.0 of both and found I could no longer build any binary targets. The problem is because the import path for packages is incorrect, so the compiler complains that a particular type cannot be used as another one.

```
cannot use conn (type *"github.com/lightstep/vendor/google.golang.org/grpc".ClientConn) as type *"google.golang.org/grpc".ClientConn in argument to iam.NewIAMPolicyClient
```

I downgraded to rules_go 0.12.1 and gazelle 0.12.0 and had no issues.

The difference I found is the importmap that's generated by gazelle:

0.13.0:
```
    importmap = "github.com/lightstep/vendor/cloud.google.com/go/civil",
```

0.12.0:
```
    importmap = "vendor/cloud.google.com/go/civil",
```

For reference, WORKSPACE is at github.com/lightstep.

This is my root BUILD:

```
gazelle(
    name = "gazelle",
    external = "vendored",
    prefix = "github.com/lightstep",
)
```

**Comment by krishicks** [comment]
When looking into this further I discovered https://github.com/bazelbuild/rules_go/issues/1450, where @jayconrod mentions using `# gazelle:importmap_prefix vendor` in vendor/BUILD.

Doing that gives me the same importmap as above, but building fails with the same sort of error as before.

The only other changes I see between 0.12.0 and 0.13.0 (with importmap_prefix set) are like this:

```diff
diff --git a/go/src/github.com/lightstep/vendor/cloud.google.com/go/iam/BUILD b/go/src/github.com/lightstep/ve
ndor/cloud.google.com/go/iam/BUILD
index a60c10f43..78afcba35 100644
--- a/go/src/github.com/lightstep/vendor/cloud.google.com/go/iam/BUILD
+++ b/go/src/github.com/lightstep/vendor/cloud.google.com/go/iam/BUILD
@@ -8,7 +8,7 @@ go_library(
     visibility = ["//visibility:public"],
     deps = [
         "//vendor/golang.org/x/net/context:go_default_library",
+        "//vendor/google.golang.org/genproto/googleapis/iam/v1:go_default_library",
         "//vendor/google.golang.org/grpc:go_default_library",
-        "@go_googleapis//google/iam/v1:iam_go_proto",
     ],
 )
```

**Comment by chrismcbride** [comment]
I believe Im having a similar problem. I now get compile errors like this when using the heroku docker registry client (imported via the go_repository rule):

`
cannot use layer.SHA256Digest (type "github.com/opencontainers/go-digest".Digest) as type "github.com/heroku/docker-registry-client/vendor/github.com/opencontainers/go-digest".Digest in argument to r.registry.HasLayer
`

The BUILD file gazelle generates looks like this:

`
go_library(
    name = "go_default_library",
    srcs = [
        "authchallenge.go",
        "basictransport.go",
        "errortransport.go",
        "json.go",
        "layer.go",
        "manifest.go",
        "registry.go",
        "repositories.go",
        "tags.go",
        "tokentransport.go",
    ],
    importpath = "github.com/heroku/docker-registry-client/registry",
    visibility = ["//visibility:public"],
    deps = [
        "//vendor/github.com/docker/distribution:go_default_library",
        "//vendor/github.com/docker/distribution/manifest/schema1:go_default_library",
        "//vendor/github.com/docker/distribution/manifest/schema2:go_default_library",
        "//vendor/github.com/opencontainers/go-digest:go_default_library",
    ],
)
`
So when I go to use the client with a dependency of @com_github_opencontainers_go_digest//:go_default_library it treats it as a separate import. Setting the build_external flag to external or vendoered makes no difference

**Comment by chrismcbride** [comment]
To add more context, this is the generated BUILD.bazel for opencontainers/go-digest:

```
load("@io_bazel_rules_go//go:def.bzl", "go_library")

go_library(
    name = "go_default_library",
    srcs = [
        "algorithm.go",
        "digest.go",
        "digester.go",
        "doc.go",
        "verifiers.go",
    ],
    importmap = "github.com/heroku/docker-registry-client/vendor/github.com/opencontainers/go-digest",
    importpath = "github.com/opencontainers/go-digest",
    visibility = ["//visibility:public"],
)
```

**Comment by jayconrod** [comment]
I commented on #300, but I'll leave a comment here as well to explain why this is happening.

If you import an external repository using `go_repository`, and that repository has a vendor directory, Gazelle will resolve imports within that repository to libraries in the vendor directory if they exist. Gazelle will not resolve any imports outside of the repository into that vendor directory because vendor directories are private according to "go build" semantics, and Gazelle's charter is to map "go build" semantics to Bazel. If types from vendored packages are exposed in repository's public API, that can cause compile-time and link-time conflicts when you bring in your own copy of the same package.

At some point (I think in rules_go 0.11.0 or 0.12.0), I introduced the `importmap` attribute and fixed how multiple copies of libraries were handled in rules_go. Prior to that, we were only compiling against one copy and ignoring other copies, which produced incorrect builds without errors. Since we weren't reporting errors for these conflicts before, a lot of people upgraded from 0.10 and saw them from the first time. In most cases I've seen though, these errors have been an issue with the build itself, rather than rules_go or Gazelle.

**Comment by jayconrod** [comment]
@krishicks Sorry for not responding to this much earlier. At this point, I don't have enough information to reproduce your issue. You mentioned there was an example repo I could try out, but `github.com/lightstep` is an organization with many repositories, so please point me to the specific one that has the issue.

**Comment by krishicks** [comment]
@jayconrod here's a reproducible example:

```sh
go get cloud.google.com/go
cd $GOPATH/src/cloud.google.com/go
dep init
dep ensure

cat > BUILD <<EOF
load("@bazel_gazelle//:def.bzl", "gazelle")
gazelle(
    name = "gazelle",
    external = "vendored",
    prefix = "cloud.google.com/go",
)
EOF

cat > WORKSPACE <<EOF
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
http_archive(
    name = "io_bazel_rules_go",
    url = "https://github.com/bazelbuild/rules_go/releases/download/0.15.3/rules_go-0.15.3.tar.gz",
    sha256 = "97cf62bdef33519412167fd1e4b0810a318a7c234f5f8dc4f53e2da86241c492",
)
load("@io_bazel_rules_go//go:def.bzl", "go_rules_dependencies", "go_register_toolchains")
go_rules_dependencies()
go_register_toolchains(go_version = "1.11")
http_archive(
    name = "bazel_gazelle",
    urls = ["https://github.com/bazelbuild/bazel-gazelle/releases/download/0.14.0/bazel-gazelle-0.14.0.tar.gz"],
    sha256 = "c0a5739d12c6d05b6c1ad56f2200cb0b57c5a70e03ebd2f7b87ce88cabf09c7b",
)
load("@bazel_gazelle//:deps.bzl", "gazelle_dependencies")
gazelle_dependencies()
EOF

bazel run //:gazelle
bazel build //iam:go_default_library
```

This results in an error similar to what I first supplied:

```log
INFO: Analysed target //iam:go_default_library (41 packages loaded).
INFO: Found 1 target...
ERROR: /home/hicks/go/src/cloud.google.com/go/iam/BUILD:3:1: GoCompile iam/linux_amd64_stripped/go_default_library%/cloud.google.com/go/iam.a failed (Exit 1)
/home/hicks/.cache/bazel/_bazel_hicks/b6424f4ba800632a23daf6f922bee732/sandbox/processwrapper-sandbox/1/execroot/__main__/iam/iam.go:107:58: cannot use conn (type *"cloud.google.com/go/vendor/google.golang.org/grpc".ClientConn) as type *"google.golang.org/grpc".ClientConn in argument to iam.NewIAMPolicyClient
GoCompile: error running subcommand: exit status 2
Target //iam:go_default_library failed to build
Use --verbose_failures to see the command lines of failed build steps.
```

Let me know if that doesn't work the same for you.

**Comment by jayconrod** [comment]
@krishicks Got it, thanks.

I think this would be resolved by fixing #217. When proto rule generation is not disabled, Gazelle should probably resolve dependencies on `google.golang.org/grpc` to `@org_golang_google_grpc//:go_default_library`. `go_proto_library` rules with the `go_grpc` compiler depend on that implicitly, and linking in a vendored copy of that package causes this error.

**Comment by krishicks** [comment]
@jayconrod I decided to re-run the above example but with rules_go 16.5 and gazelle 0.16.0, and do not see the same failure as above.

I'll close this issue.

Thanks!
